source/MRViewer/MRSurfaceContoursWidget.cpp

```diff
@@ -0,0 +1,421 @@
+        return false;
+
+    auto [obj, pick] = getViewerInstance().viewport().pick_render_object();
+
+    auto addPoint = [this] ( const std::shared_ptr<ObjectMeshHolder> obj, const MeshTriPoint& triPoint, bool close )
+    {
+        if ( !pickedPoints_[obj].empty() )
+            updateBaseColor( pickedPoints_[obj].back(), Color::gray() );
+
+        AppendHistory<AddPointActionPickerPoint>( *this, obj, triPoint );
+        pickedPoints_[obj].push_back( createPickWidget_( obj, triPoint ) );
+        updateBaseColor( pickedPoints_[obj].back(), close ? Color::transparent() : Color::green() );
+        onPointAdd_( obj );
+    };
+    auto removePoint = [this] ( const std::shared_ptr<ObjectMeshHolder> obj, int pickedIndex )
+    {
+        if ( pickedIndex == int( pickedPoints_[obj].size() ) - 1 && pickedPoints_[obj].size() > 1 )
+        {
+            SurfacePointWidget::Parameters params;
+            params.baseColor = Color::green();
+            pickedPoints_[obj][pickedIndex - 1]->setParameters( params );
+        }
+
+        AppendHistory<RemovePointActionPickerPoint>( *this, obj, pickedPoints_[obj][pickedIndex]->getCurrentPosition(), pickedIndex );
+        pickedPoints_[obj].erase( pickedPoints_[obj].begin() + pickedIndex );
+        activeIndex_ = pickedIndex;
+        activeObject_ = obj;
+        onPointRemove_( obj );
+    };
+
+    if ( !mod ) // just add new point 
+    {
+        auto objMesh = std::dynamic_pointer_cast< ObjectMeshHolder >( obj );
+        if ( !objMesh )
+            return false;
+
+        // all pick in point (without mod) must not comes here. Point widget must "eat" them. 
+        if ( isClosedCountour( objMesh ) )
+            return false;
+
+        assert( objMesh != nullptr ); // contoursWidget_ can join for mesh objects only
+
+        auto triPoint = objMesh->mesh()->toTriPoint( pick );
+        addPoint( objMesh, triPoint, false );
+        return true;
+    }
+    else if ( mod == GLFW_MOD_CONTROL ) // close contour case 
+    {
+        // Try to find parent object 
+        auto isFirstPointOnCountourClicked = false;
+        std::shared_ptr<ObjectMeshHolder> objectToCloseCoutour = nullptr;
+        for ( const auto& [parentObj, contour] : pickedPoints_ )
+        {
+            if ( contour.size() > 2 && obj == contour[0]->getPickSphere() )
+            {
+                isFirstPointOnCountourClicked = true;
+                objectToCloseCoutour = parentObj;
+                break;
+            }
+        }
+        if ( !isFirstPointOnCountourClicked )
+            return false;
+
+        if ( isClosedCountour( objectToCloseCoutour ) )
+            return false;
+
+        assert( objectToCloseCoutour != nullptr );
+        auto triPoint = pickedPoints_[objectToCloseCoutour][0]->getCurrentPosition();
+        addPoint( objectToCloseCoutour, triPoint, true );
+        return true;
+    }
+    else if ( mod == GLFW_MOD_SHIFT )  // remove point case 
```

Comment:
should it be configurable with some parameters?

Reply:
done